wallet-new/src/Cardano/Wallet/Kernel/BListener.hs

```diff
@@ -63,7 +68,32 @@ applyBlock pw@PassiveWallet{..} b = do
       Right confirmed -> do
         modifyMVar_ _walletSubmission $ return . Submission.remPending confirmed
         mapM_ (putTxMeta _walletMeta) metas
+        createCheckpointIfNeeded
         return $ Right ()
+  where
+      -- | Interim fix, see CBR-438 and
+      -- https://github.com/acid-state/acid-state/issues/103. In brief, when
+      -- the note initially syncs and lots of blocks gets passed to the wallet
+      -- worker, a new `ApplyBlock` acidic transaction will be committed on the
+      -- transaction log but not written to disk _yet_ (that's what checkpoints
+      -- are for). However, this might lead to memory leaks if such checkpointing
+      -- step doesn't happen fast enough. Therefore, every time we apply a block,
+      -- we decrement this counter and when it reaches 0, we enforce a new
+      -- checkpoint.
+      createCheckpointIfNeeded :: IO ()
+      createCheckpointIfNeeded = do
+          -- Look at the 'ResolvedBlock' 's 'SlotId', and assess if a new
+          -- checkpoint is needed by doing @localBlockIx `modulo` someConstant@
+          -- where @someConstant@ is chosen to be 1000.
+          let blockSlotIx = b ^. rbContext
+                               . bcSlotId
+                               . fromDb
+                               . siSlotL
+                               . to getSlotIndex
+              checkpointNeeded = (blockSlotIx - 1) `mod` 1000 == 0
```

Comment:
This was done to avoid situations where skipping block 0 would have not yielded the requested checkpoint. Note that even if we hit `0 - 1` that would result in an underflow (`getSlotIndex` returns a `Word16`) this would simply mean our `mod` won't yield 0 and we would skip the checkpoint, but no catastrophic behaviour would incur.